cmd/kubeadm/app/cmd/join.go

```diff
@@ -245,27 +296,181 @@ func NewJoin(cfgPath string, args []string, defaultcfg *kubeadmapiv1alpha3.JoinC
+		preflight.RunInitMasterChecks(utilsexec.New(), masterConfiguration, j.ignorePreflightErrors)
+
+		// Prepares the node for joining as a master by writing necessary
+		// KubeConfig files, and static pod manifests
+		if err = j.PrepareForJoinAsMaster(masterConfiguration); err != nil {
+			return err
+		}
+	}
+
+	// Executes the kubelet TLS bootstrap process, that completes with the node
+	// joining the cluster with a dedicates set of credentials as required by
+	// the node authorizer.
+	// if the node is joining as a master, since it uses static pods for the control plane,
+	// as soon as the kubelet starts it will take charge of creating control plane
+	// components on the node.
+	if err = j.BootstrapKubelet(tlsBootstrapCfg); err != nil {
+		return err
+	}
+
+	// if the node is joining as a master
+	if j.cfg.Master == true {
+		// Marks the node with master taint and label.
+		if err := j.MarkMaster(); err != nil {
+			return err
+		}
+
+		// outputs the join master done template and exits
+		ctx := map[string]string{
+			"KubeConfigPath": kubeadmconstants.GetAdminKubeConfigPath(),
+		}
+		joinMasterNodeDoneTemp.Execute(out, ctx)
+		return nil
+	}
+
+	// otherwise, if the node joined as a worker node;
+	// outputs the join done message and exits
+	fmt.Fprintf(out, joinWorkerNodeDoneMsg)
+	return nil
+}
+
+// FetchInitClusterConfiguration reads the master configuration from the kubeadm-admin configMap,
+func (j *Join) FetchInitClusterConfiguration(tlsBootstrapCfg *clientcmdapi.Config) (*kubeadmapi.InitConfiguration, error) {
+	// creates a client to access the cluster using the bootstrap token identity
+	tlsClient, err := kubeconfigutil.ToClientSet(tlsBootstrapCfg)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to access the cluster: %v", err)
+	}
+
+	// Fetches the master configuration from the cluster
+	kubeadmConfig, err := configutil.FetchConfigFromFileOrCluster(tlsClient, os.Stdout, "join", "")
 	if err != nil {
+		return nil, fmt.Errorf("Unable to fetch the kubeadm-config ConfigMap: %v", err)
+	}
+
+	// Converts public API struct to internal API
+	masterConfiguration := &kubeadmapi.InitConfiguration{}
+	kubeadmscheme.Scheme.Convert(kubeadmConfig, masterConfiguration, nil)
+
+	return masterConfiguration, nil
+}
+
+// CheckIfReadyForJoinAsMaster ensures that the cluster is in a state that supports
+// joining an additional master node and if the node is ready to join
+func (j *Join) CheckIfReadyForJoinAsMaster(masterConfiguration *kubeadmapi.InitConfiguration) error {
+	// blocks if the cluster was created without a stable controlplane endpoint
+	if masterConfiguration.API.ControlPlaneEndpoint == "" {
+		return fmt.Errorf("unable to join a new master node on a cluster that doesn't have a stable api.controlPlaneEndpoint address")
+	}
+
+	// blocks if the cluster was created without an external etcd cluster
+	if masterConfiguration.Etcd.External == nil {
```

Comment:
Could you wrap this with a //TODO and open an issue. 
In an ideal work we bootstrap etcd with the join scenario as well.

Reply:
Added as a point in the tracking issue